README.md
  
[![Build Status](https://travis-ci.org/mjirik/anwa.svg?branch=master)](https://travis-ci.org/mjirik/anwa)
[![Coverage Status](https://coveralls.io/repos/github/mjirik/anwa/badge.svg?branch=master)](https://coveralls.io/github/mjirik/anwa?branch=master)
[![PyPI version](https://badge.fury.io/py/anwa.svg)](http://badge.fury.io/py/anwa)


# anwa

Automatic animal detection in video


# Install

* Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) and during install add conda to path

* In command line (`cmd` on Windows) use:

```cmd
    conda create -n anwa_app -c conda-forge -c mjirik anwa
    conda activate anwa_app
    conda -m anwa install
```


changelog.md
# Changelog

## 0.1.0

* [added] Cut videos based on activity

tests/anwa_test.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

# import logging
# logger = logging.getLogger(__name__)
from loguru import logger
import pytest
import os.path
import anwa.algorithm

path_to_script = os.path.dirname(os.path.abspath(__file__))



# @pytest.mark.interactive
# @pytest.mark.slow
def test_answer():
    aw = anwa.algorithm.AnimalWatch()
    aw.set_input_dir("~/data/lynx_lynx/fotopasti_20170825/videa/bez rysa/lok3/2017_01_17/")
    # aw.set_input_dir("~/data/lynx_lynx/fotopasti_20170825/videa/s rysem/**/*")
    aw.run()




anwa/activity_detector.py
import glob
import os.path as op
# import seaborn as sns
from pyqtgraph.parametertree import Parameter
import imageio
import skimage
import numpy as np
import datetime
from pathlib import Path
import matplotlib.pyplot as plt
from loguru import logger
from exsu.report import Report



# datapath = op.expanduser("~/data/lynx_lynx/fotopasti_20170825/videa/s rysem/**/**")


class ActivityDetector:
    def __init__(self, report=None):
        params = [
            {
                "name": "Activity Video Cut Threshold",
                "type": "float",
                "value": 1.0,
                'tip': "Applied on activity frames cropping out frames without activity",
                "suffix": "%",
                "siPrefix": False

            },
            {
                "name": "Activity Crop Threshold",
                "type": "float",
                "value": .3,
                'tip': "Applied on activity frames filtered by time and space"
                # "suffix": "",
                # "siPrefix": True

            },
            {
                "name": "Start Frame Offset",
                "type": "int",
                "value": 1,
                'tip': "ignore first few frames"
                # "suffix": "",
                # "siPrefix": True

            },
            {
                "name": "End Frame Offset",
                "type": "int",
                "value": 0,
                'tip': "ignore last few frames"
                # "suffix": "",
                # "siPrefix": True

            },
            # {
            #     "name": "GLCM Levels",
            #     "type": "int",
            #     "value": 64
            # },

        ]

        self.parameters = Parameter.create(name="Activity Detector", type="group", children=params, expanded=False)
        self.input_path = None
        self.output_path = None
        self.report=report

    def set_input_path(self, input_path):
        """
        :param input_path: can be with '~' for home dir and with /**/* for any subdir
        :return:
        """
        self.input_path = Path(input_path)

    def set_output_path(self, output_path):
        self.output_path = Path(output_path)


    def run(self):
        logger.debug(f"Reading files in dir: {self.input_path}")
        # fnvideos = glob.glob(op.join(op.expanduser(self.input_path), "*"))
        fnvideos = list(Path(op.join(op.expanduser(self.input_path))).glob('**/*'))
        logger.debug(f"number files in dir: {len(fnvideos)}")
        for fn in fnvideos:
            plt.figure(figsize=(14, 8))
            try:
                self.run_one(Path(fn))
            # except Exception as e:
            except ValueError as e:
                import traceback
                logger.warning(f"Problem with processing file: {fn}")
                logger.debug(traceback.format_exc())
        pass

    def run_one(self, fn:Path):
        """
        """
        report = self.report
        logger.info(f"reading file {fn}")
        vid = imageio.get_reader(fn)
        logger.debug("background model extraction")

        background_model, frame_size = create_background_model(vid)
        if report is not None:
            logger.debug("saving")
            logger.debug(
                f"min: {np.min(background_model)}, mean: {np.mean(background_model)}, max: {np.max(background_model)}")
            report.imsave(f"{fn.stem}_background", background_model / 255.0, level=50, level_skimage=10)
        logger.debug("activity estimation")
        begin_offset = int(self.parameters.param("Start Frame Offset").value())
        end_offset = int(self.parameters.param("End Frame Offset").value())
        # end_offset = 10
        error, time, errmax = activity_estimation(vid, background_model, begin_offset=begin_offset, end_offset=end_offset)
        import scipy
        error_filt = scipy.signal.medfilt(error, 5)
        if report is not None:
            fig = plt.figure()
            self.plt_activity(time, error, error_filt=error_filt)
            report.savefig_and_show(fn.stem + "_activity", fig)
        frame, iframe = get_max_image(vid, error)
        if report is not None:
            report.imsave(f"{fn.stem}_maxframe", frame, level=50, level_skimage=10)
        errim = get_activity_diff_image(frame, background_model)
        if report is not None:
            report.imsave(f"{fn.stem}_errim", errim, level=45, level_skimage=10)
        #     return background_model, error, time, errmax, image, imax
        filtered_activity = activity_filter_time_and_space(vid, iframe, background_model)
        if report is not None:

            report.imsave(f"{fn.stem}_filtered_activity", filtered_activity, level=40, level_skimage=10)
            fig = plt.figure()
            plt.imshow(filtered_activity)
            plt.colorbar()
            report.savefig_and_show(f"{fn.stem}_filtered_activity_fig", fig, level=40)

        thr = float(self.parameters.param("Activity Crop Threshold").value())
        roi = activity_roi(filtered_activity, activity_threshold=thr)
        if roi is not None:
            cropped_frame = crop_frame(frame, *roi)
            if report is not None:
                report.imsave(f"{fn.stem}_max_activity_crop", cropped_frame, level=60, level_skimage=10)

        self._cut_video(vid, fn, error_filt)


    def _cut_video(self, vid, fn:Path, error):

        print("cut video")
        thr = float(self.parameters.param("Activity Video Cut Threshold").value()) * 0.01
        reader = vid
        # reader = imageio.get_reader('imageio:cockatoo.mp4')
        fps = reader.get_meta_data()['fps']
        # import pdb; pdb.set_trace()
        fnrel = fn.relative_to(self.input_path.expanduser())
        logger.debug(f"relative path {str(fnrel)}")

        opath = self.output_path/fnrel.parent/(fnrel.stem + "_cut" + fn.suffix)
        logger.debug(f"opath {str(opath)}")
        opath.parent.mkdir(parents=True, exist_ok=True)

        writer = imageio.get_writer(opath, fps=fps)
        logger.info(f"write video: {str(opath)}")
        logger.debug(f"video cut threshold: {thr} ")

        i = int(self.parameters.param("Start Frame Offset").value())
        for im in reader:
            if i < len(error):
                if error[i] > thr:
                    writer.append_data(im)  # [:, :, 1])
                    # pass
            i += 1
        writer.close()

    def plt_activity(self, time, error, error_filt=None):
        thr_percent = float(self.parameters.param("Activity Video Cut Threshold").value())
        plt.semilogy(time, 100 * (error))
        if error_filt is not None:
            plt.semilogy(time, 100 * (error_filt), "r")
        # plt.ylim(base, 1+base)
        plt.ylim(0.1, 100)
        # plt.hlines([thr_percent])
        logger.debug(f"thr percent: {thr_percent}")
        plt.axhline(thr_percent, ls='--')
        plt.xlabel("Time [s]")
        plt.ylabel("Activity [%]")




def create_background_model(vid, begin_offset=1, end_offset=50, step=10):

    vid_len = vid.get_length() - end_offset - begin_offset
    frame_size = vid.get_meta_data()["size"]

    background_model = np.zeros([frame_size[1], frame_size[0], 3])

    # nums = range(int(begin_offset), int(vid.get_length() - end_offset), int(step))
    nums = range(begin_offset, vid.count_frames() - end_offset, step)

    # nums = [1,5, 10]

    processed_images = 0
    for num in nums:
        image = vid.get_data(num).astype(np.float)
        if processed_images == 0:
            background_model = image
        else:
            background_model = (processed_images * background_model) / (processed_images + 1.) + image / (
                        processed_images + 1.)
        processed_images += 1

    return background_model, frame_size


def activity_estimation(vid, background_model, begin_offset=0, end_offset=0, step=1):
    # error = np.zeros([vid.count_frames()])
    # error = [0.0] * vid.count_frames()
    error = []

    fps = vid.get_meta_data()['fps']
    frame_size = vid.get_meta_data()["size"]
    nums = range(begin_offset, vid.count_frames() - end_offset, step)

    # times = []
    errmax = 255 * 3 * np.prod(frame_size)
    #     print(errmax)
    for num in nums:
        #     print(num)
        image = vid.get_data(num).astype(np.float)
        err = np.sum(np.abs(background_model - image))
        # error[num] = err
        error.append(err)

    time = np.asarray(nums) / fps
    error = np.array(error) / errmax
    return error, time, errmax




import skimage.filters
from skimage.morphology import disk


def get_max_image(vid, error, begin_offset=1, end_offset=10):
    import copy
    error2 = copy.copy(error)
    error2[-end_offset:] = 0
    imax = np.argmax(error2)
    image = get_image(vid, imax, begin_offset)
    return image, imax


def get_image(vid, imax, begin_offset=1):
    return vid.get_data(imax + begin_offset)



def get_activity_diff_image(image, background_model, safety_multiplicator=0.99):
    diff = np.abs(image.astype(float) - background_model.astype(float))
    errim = (diff[:, :, 0] ** 2 + diff[:, :, 1] ** 2 + diff[:, :, 2] ** 2) ** 0.5
    # normaliza to range <0 ; 1>
    # sqrt(3) because of 3 color channels
    errim = safety_multiplicator * errim / (255. * np.sqrt(3))
    return errim


def activity_time_filter(
        vid, iframe, background_model,
        begin_offset=1, end_offset=50, around_imax=range, time_range=5):
    around_mn = int(max(begin_offset, round(iframe - (time_range / 2))))
    around_mx = int(min(vid.count_frames() - end_offset, round(iframe + (time_range / 2))))

    around_range = range(around_mn, around_mx, 10)

    frames = np.zeros([background_model.shape[0], background_model.shape[1], len(around_range)])

    for i, iiframe in enumerate(around_range):
        #     print(i, iframe)
        frames[:, :, i] = get_activity_diff_image(get_image(vid, iiframe, begin_offset=begin_offset), background_model)
    #     plt.figure()
    #     plt.imshow(frames[:,:,i])

    time_mean_frame = np.mean(frames, axis=2)
    return time_mean_frame


def activity_space_filter(time_mean_frame, median_disk_radius=15, gaussian_sigma=5.0):
    mn = np.min(time_mean_frame)
    print("limits ", mn, np.max(time_mean_frame))
    if mn < -1:
        time_mean_frame = time_mean_frame - mn
    #     if np.min(time_mean_frame) < 0:

    # mean_frame_filtered = skimage.filters.gaussian(mean_frame, sigma=5.0)
    mean_frame_filtered = skimage.filters.median(time_mean_frame, disk(median_disk_radius))
    mean_frame_filtered = skimage.filters.gaussian(mean_frame_filtered, sigma=gaussian_sigma)
    return mean_frame_filtered


import matplotlib.patches as patches

def activity_filter_time_and_space(vid, iframe, background_model):
    """
    Focus on few frames around maximum intensity frame
    :param vid:
    :param iframe:
    :param background_model:
    :return:
    """
    im = activity_time_filter(vid, iframe, background_model)
    im = activity_space_filter(im)
    return im


def roi_size(roi_min, roi_max):
    roi_y_size = roi_max[0] - roi_min[0]
    roi_x_size = roi_max[1] - roi_min[1]
    roi_size = [roi_y_size, roi_x_size]
    return roi_size


def activity_roi(mean_frame_filtered, cut_border=40, activity_threshold=0.1):
    binary_activity_frame = mean_frame_filtered > activity_threshold
    x, y = np.nonzero(binary_activity_frame)
    if len(x) == 0:
        return None
    roi_min = [
        max(np.min(y) - cut_border, 0),
        max(np.min(x) - cut_border, 0)
    ]
    roi_max = [
        min(np.max(y) + cut_border, mean_frame_filtered.shape[1]),
        min(np.max(x) + cut_border, mean_frame_filtered.shape[0]),

    ]
    return roi_min, roi_max

def show_roi(roi_min, roi_max=None, roi_size=None, image=None):
    if roi_size is None:
        roi_y_size = roi_max[0] - roi_min[0]
        roi_x_size = roi_max[1] - roi_min[1]
        roi_size = [roi_y_size, roi_x_size]
    ax = plt.gca()
    if image is not None:
        plt.imshow(image)
    # plt.imshow(binary_activity_frame)

    ax.add_patch(
        patches.Rectangle(
            roi_min,
    #         (10, 10),
            roi_size[0],
            roi_size[1],
            fill=False,      # remove background
            edgecolor="red",
            linewidth=3
        )
    )

def crop_frame(frame, roi_min, roi_max):
    return frame[
        roi_min[1]:roi_max[1],
        roi_min[0]:roi_max[0],
        :
    ]




anwa/algorithm.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
"""
Modul is used for GUI of Lisa
"""

from loguru import logger

# from scaffan import image
from pathlib import Path
import sys
import os.path as op
import datetime
# import io3d.misc
from io3d import cachefile
import json
import time
import platform

# import PyQt5.QtWidgets
# print("start 3")
# from PyQt5.QtWidgets import QApplication, QFileDialog
# print("start 4")
from PyQt5 import QtGui

# print("start 5")
from pyqtgraph.parametertree import Parameter, ParameterTree
import pyqtgraph.widgets
import pyqtgraph.widgets

# print("start 6")

import io3d
import io3d.datasets
pth = str(Path(__file__).parent / "../../exsu/")
logger.debug(f"exsu path: {pth}")
sys.path.insert(0, pth)
import exsu
logger.debug(f"exsu path 2: {exsu.__file__}")
from exsu.report import Report

# import scaffan.lobulus
# import scaffan.report
# import scaffan.skeleton_analysis
# from .report import Report
# import scaffan.evaluation
# from scaffan.pyqt_widgets import BatchFileProcessingParameter
from . import activity_detector


class AnimalWatch:
    def __init__(self):

        self.report: exsu.report.Report = exsu.report.Report()
        self.report.set_save(True)
        # self.report.level = 50

        self.raise_exception_if_color_not_found = True

        # import scaffan.texture as satex
        # self.glcm_textures = satex.GLCMTextureMeasurement()
        # self.lobulus_processing = scaffan.lobulus.Lobulus()
        # self.skeleton_analysis = scaffan.skeleton_analysis.SkeletonAnalysis()
        # self.evaluation = scaffan.evaluation.Evaluation()

        # self.lobulus_processing.set_report(self.report)
        # self.glcm_textures.set_report(self.report)
        # self.skeleton_analysis.set_report(self.report)
        # self.evaluation.report = self.report
        self.activity_detector:activity_detector.ActivityDetector = activity_detector.ActivityDetector(
            report=self.report)
        self.win:QtGui.QWidget = None
        self.cache = cachefile.CacheFile("~/.exsu_cache.yaml")
        # self.cache.update('', path)
        common_spreadsheet_file = self.cache.get_or_save_default(
            "common_spreadsheet_file", self._prepare_default_output_common_spreadsheet_file())
        logger.debug("common_spreadsheet_file loaded as: {}".format(common_spreadsheet_file))
        params = [
            {
                "name": "Input",
                "type": "group",
                "children": [
                    {"name": "Dir Path", "type": "str"},
                    {"name": "Select", "type": "action"},
                    {"name": "Data Info", "type": "str", "readonly": True},
                    {"name": "Camera ID", "type": "str"}
                ],
            },
            {
                "name": "Output",
                "type": "group",
                "children": [
                    {
                        "name": "Directory Path",
                        "type": "str",
                        "value": self._prepare_default_output_dir(),
                    },
                    {"name": "Select", "type": "action"},
                    # {
                    #     "name": "Common Spreadsheet File",
                    #     "type": "str",
                    #     "value": common_spreadsheet_file,
                    # },
                    # {"name": "Select Common Spreadsheet File", "type": "action",
                    #  "tip": "All measurements are appended to this file in addition to data stored in Output Directory Path."
                    #  },
                ],
            },
            {
                "name": "Processing",
                "type": "group",
                "children": [
                    # {'name': 'Directory Path', 'type': 'str', 'value': prepare_default_output_dir()},
                    {
                        'name': "Anwa dir",
                        "type": 'str',
                        'value': self._prepare_anwa_dir()
                    },
                    self.activity_detector.parameters,
                    # {
                    #     "name": "Show",
                    #     "type": "bool",
                    #     "value": False,
                    #     "tip": "Show images",
                    # },
                    # {
                    #     "name": "Open output dir",
                    #     "type": "bool",
                    #     "value": False,
                    #     "tip": "Open system window with output dir when processing is finished",
                    # },
                    # {
                    #     "name": "Run Skeleton Analysis",
                    #     "type": "bool",
                    #     "value": True,
                    #     # "tip": "Show images",
                    # },
                    # {
                    #     "name": "Run Texture Analysis",
                    #     "type": "bool",
                    #     "value": True,
                    #     # "tip": "Show images",
                    # },
                    # {
                    #     "name": "Report Level",
                    #     "type": "int",
                    #     "value": 50,
                    #     "tip": "Control ammount of stored images. 0 - all debug imagess will be stored. "
                    #            "100 - just important images will be saved.",
                    # },
                ],
            },
            {"name": "Run", "type": "action"},
        ]
        self.parameters = Parameter.create(name="params", type="group", children=params)
        # here is everything what should work with or without GUI
        # self.parameters.param("Input", "File Path").sigValueChanged.connect(
        #     self._get_file_info
        # )
        # self.anim: image.AnnotatedImage = None
        pass

    def _prepare_default_output_common_spreadsheet_file(self):
        pass

    def _prepare_default_output_dir(self):
        default_dir = io3d.datasets.join_path("lynx_lynx", "processed", "anwa", get_root=True)
        return default_dir

    def _prepare_anwa_dir(self):
        anwa_dir = Path("~/.anwa")
        return anwa_dir

    def select_input_dir_gui(self):
        from PyQt5 import QtWidgets

        default_dir = io3d.datasets.join_path(get_root=True)
        # default_dir = op.expanduser("~/data")
        if not op.exists(default_dir):
            default_dir = op.expanduser("~")

        # dialog = QtWidgets.QFileDialog()
        # dialog.setFileMode(QtWidgets.QFileDialog.Directory)
        # dialog.exec();
        fn = QtWidgets.QFileDialog.getExistingDirectory(
            None,
            "Select Input Directory",
            directory=default_dir,
            # filter="NanoZoomer Digital Pathology Image(*.ndpi)",
        )
        self.set_input_dir(fn)

    def set_input_dir(self, fn):
        fn = str(fn)
        fnparam = self.parameters.param("Input", "Dir Path")
        fnparam.setValue(fn)
        logger.debug("Set Input File Path to : {}".format(fn))
        # import pdb; pdb.set_trace()
        # print("ahoj")

    def set_output_dir(self, path):
        fnparam = self.parameters.param("Output", "Directory Path")
        fnparam.setValue(str(path))

    def set_report_level(self, level:int):
        fnparam = self.parameters.param("Processing", "Report Level")
        fnparam.setValue(level)

    def set_common_spreadsheet_file(self, path):
        fnparam = self.parameters.param("Output", "Common Spreadsheet File")
        fnparam.setValue(path)
        self.cache.update('common_spreadsheet_file', path)
        logger.debug("common_spreadsheet_file set to {}".format(path))
        print("common_spreadsheet_file set to {}".format(path))

    def select_output_dir_gui(self):
        from PyQt5 import QtWidgets

        default_dir = self._prepare_default_output_dir()
        if op.exists(default_dir):
            start_dir = default_dir
        else:
            start_dir = op.dirname(default_dir)

        fn = QtWidgets.QFileDialog.getExistingDirectory(
            None,
            "Select Output Directory",
            directory=start_dir,
            # filter='All Files(*.*)',
            # filter="NanoZoomer Digital Pathology Image(*.ndpi)"
        )
        # print (fn)
        self.set_output_dir(fn)

    def run(self):

        odir = str(self.parameters.param("Output", "Directory Path").value())
         # = str(self.parameters.param("Output", "Directory Path").value())
        self.report.level = 40

        logger.debug(f"output dir {odir}")
        self.report.init_with_output_dir(odir)
        # self.report.outputdir = odir
        Path(odir).mkdir(parents=True, exist_ok=True)
        self.activity_detector.run()
        pass

    def start_gui(self, skip_exec=False, qapp=None):

        from PyQt5 import QtWidgets
        import anwa.qtexceptionhook

        # import QApplication, QFileDialog
        if not skip_exec and qapp == None:
            qapp = QtWidgets.QApplication(sys.argv)

        self.parameters.param("Input", "Select").sigActivated.connect(
            self.select_input_dir_gui
        )
        self.parameters.param("Output", "Select").sigActivated.connect(
            self.select_output_dir_gui
        )
        # self.parameters.param("Output", "Select Common Spreadsheet File").sigActivated.connect(
        #     self.select_output_spreadsheet_gui
        # )
        self.parameters.param("Run").sigActivated.connect(
            self.run
        )

        # self.parameters.param("Processing", "Open output dir").setValue(True)
        t = ParameterTree()
        t.setParameters(self.parameters, showTop=False)
        # t.setWindowTitle("pyqtgraph example: Parameter Tree")
        # t.show()

        # print("run scaffan")
        win = QtGui.QWidget()
        win.setWindowTitle("AnWa {}".format(anwa.__version__))
        logo_fn = op.join(op.dirname(__file__), "anwa.png")
        app_icon = QtGui.QIcon()
        # app_icon.addFile(logo_fn, QtCore.QSize(16, 16))
        app_icon.addFile(logo_fn)
        win.setWindowIcon(app_icon)
        # qapp.setWindowIcon(app_icon)
        layout = QtGui.QGridLayout()
        win.setLayout(layout)
        pic = QtGui.QLabel()
        pic.setPixmap(QtGui.QPixmap(logo_fn).scaled(100, 100))
        pic.show()
        # layout.addWidget(QtGui.QLabel("These are two views of the same data. They should always display the same values."), 0,  0, 1, 2)
        layout.addWidget(pic, 1, 0, 1, 1)
        layout.addWidget(t, 2, 0, 1, 1)
        # layout.addWidget(t2, 1, 1, 1, 1)

        win.show()
        win.resize(800, 800)
        self.win = win
        # win.
        if not skip_exec:

            qapp.exec_()

